crates/cubecl-common/src/bytes/shared.rs

```diff
@@ -0,0 +1,450 @@
+    bytes: bytes::Bytes,
+    /// Lazily initialized mutable controller (copy-on-write).
+    controller: UnsafeCell<Option<Box<dyn AllocationController>>>,
+    /// Whether the controller has been initialized (data copied to heap).
+    init: AtomicBool,
+}
+
+impl SharedBytesAllocationController {
+    /// Create a new controller from a [`bytes::Bytes`] buffer.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use cubecl_common::bytes::SharedBytesAllocationController;
+    ///
+    /// // From static data (zero-copy, no heap allocation)
+    /// let static_bytes = bytes::Bytes::from_static(&[1, 2, 3, 4]);
+    /// let controller = SharedBytesAllocationController::new(static_bytes);
+    ///
+    /// // From owned data
+    /// let owned_bytes = bytes::Bytes::from(vec![1, 2, 3, 4]);
+    /// let controller = SharedBytesAllocationController::new(owned_bytes);
+    /// ```
+    pub fn new(bytes: bytes::Bytes) -> Self {
+        Self {
+            bytes,
+            controller: UnsafeCell::new(None),
+            init: AtomicBool::new(false),
+        }
+    }
+
+    /// Copy the shared bytes into a mutable native allocation controller.
+    /// This is called lazily on first mutable access (copy-on-write).
+    fn init_mutable(&self) {
+        if self.init.load(Ordering::Relaxed) {
+            return;
+        }
+
+        let data: &[u8] = &self.bytes;
+        let mut buf = vec![0u8; data.len()];
+        buf.copy_from_slice(data);
+
+        let controller = NativeAllocationController::from_elems(buf);
+        // SAFETY: We only write to the UnsafeCell when init is false,
+        // and set init to true immediately after. This is safe because
+        // UnsafeCell makes this type !Sync, preventing concurrent access.
+        // The atomic bool provides an efficient check without locking.
+        unsafe {
+            *self.controller.get() = Some(Box::new(controller));
+        }
+        self.init.store(true, Ordering::Relaxed);
+    }
+}
+
+impl AllocationController for SharedBytesAllocationController {
+    fn alloc_align(&self) -> usize {
+        if self.init.load(Ordering::Relaxed) {
+            // After copy-on-write, use the native controller's alignment.
+            // SAFETY: init is true, so controller is guaranteed to be Some.
+            unsafe {
+                (*self.controller.get())
+                    .as_ref()
+                    .expect("controller must be Some when init is true")
+                    .alloc_align()
+            }
+        } else {
+            // bytes::Bytes doesn't guarantee any particular alignment.
+            // Report byte alignment (1) since we support arbitrary offsets.
+            1
+        }
+    }
+
+    fn property(&self) -> AllocationProperty {
+        // bytes::Bytes could be backed by static data, heap, or mmap (via from_owner).
+        // We report Other since we don't know the underlying storage type.
+        AllocationProperty::Other
```

Comment:
We could pass it as parameter in the `new` function of the allocator. That can be important to speedup the loading by using pinned memory staging buffers when loading from a file.